### PR TITLE
Replace DatasetBuilder with FactoryBot

### PR DESCRIPTION
**Pull request by benthorner**
https://trello.com/c/9IACvTIA/238-replace-custom-datasetbuilder-factor-with-factorybot

FactoryBot is the standard solution for creating model objects or testing. Using FactoryBot significantly reduces the size of the tests!

Some of the tests now make more use of the defaults specified in the factory. This helps keep the test size down and encourages making the defaults realistic.

### CODE REVIEW

File: spec/features/search_page_spec.rb [hunk 1]
```diff
@@ -4,113 +4,71 @@
...
-                .build
-
+    dataset = build :dataset, title: 'a title'
     index(dataset)
-    search_for(query)
+    search_for('title')
 
     expect(page).to have_css('h1', text: 'Search results')
-    expect(page).to have_css('a', text: dataset_title)
+    expect(page).to have_css('a', text: 'title')
```
Comment (andrewgarner) [inline spec/features/search_page_spec.rb]: This assertion is now very generic – the page should have an anchor containing the text "title". There is a risk this could pass for the wrong reason.

Are we asserting it should have a link to the search term — which is now "title" — or is it a sub match on the title of the dataset — which is now "a title" — we expect to be in the results?

File: spec/features/search_page_spec.rb [hunk 2]
```diff
@@ -4,113 +4,71 @@
...
 
+    index(dataset1, dataset2)
     visit('/search')
-
     assert_data_set_length_is(2)
 
     execute_script "window.location = '#publisher'"
-
-    find_field('publisher').send_keys('Tob', :down, :enter)
+    find_field('publisher').send_keys(dataset1.organisation.title, :down, :enter)
```
Comment (andrewgarner) [inline spec/features/search_page_spec.rb]: This test is simulating "Tob" being typed, and the autocomplete behaviour matching the full name of the publisher "Toby Corp". We've not lost that intent.

File: spec/features/search_page_spec.rb (hunk 1 above)
Comment (benthorner) [reply spec/features/search_page_spec.rb]: Fair enough - being concise was my aim.

File: spec/features/search_page_spec.rb (hunk 2 above)
Comment (benthorner) [reply spec/features/search_page_spec.rb]: Good point - done.

File: spec/features/search_page_spec.rb [hunk 3]
```diff
@@ -4,63 +4,40 @@
...
-                    .with_name('old-dataset')
-                    .public_updated_at(Time.now - 1)
-                    .build
-
-    new_dataset = DatasetBuilder.new
-                    .with_title('Recent Interesting Dataset')
-                    .with_name('new-dataset')
-                    .build
-
+    old_dataset = build :dataset, title: 'Old Interesting Dataset', public_updated_at: 1.hour.ago
```
Comment (benthorner) [inline spec/features/search_page_spec.rb]: The previous difference of 1 second caused the tests to flash.

File: spec/features/search_page_spec.rb [hunk 4]
```diff
@@ -4,63 +4,40 @@
...
+    new_dataset = build :dataset, title: 'Recent Interesting Dataset'
     index(old_dataset, new_dataset)
 
     search_for('Old Interesting Dataset')
-
     expect(page).to have_css('option[selected]', text: 'Best match')
```
Comment (benthorner) [inline spec/features/search_page_spec.rb]: The large number of new lines here was making the test hard to follow - removed a few to group the together a bit more.

File: spec/features/search_page_spec.rb [hunk 5]
```diff
@@ -69,48 +46,29 @@
...
-                      .with_publisher('Tonka Trucks')
-                      .build
-
-    second_dataset = DatasetBuilder.new
-                      .with_title('Data About Toby')
-                      .with_publisher('Toby Corp')
-                      .build
-
-    index(first_dataset, second_dataset)
+    dataset1 = build :dataset, title: 'Wanted'
```
Comment (benthorner) [inline spec/features/search_page_spec.rb]: Using factory default texts should make it easier to concentrate on the purpose of the test and less on the amusing content.

File: spec/features/search_page_spec.rb [hunk 6]
```diff
@@ -121,165 +79,88 @@
...
+    dataset2 = build :dataset, :unrelated
 
+    index(dataset1, dataset2)
     visit('/search')
-
     assert_data_set_length_is(2)
 
     execute_script "window.location = '#format'"
-    find_field('format').send_keys('FO', :down, :enter)
+    find_field('format').send_keys('CSV', :down, :enter)
```
Comment (benthorner) [inline spec/features/search_page_spec.rb]: Using factory defaults makes the test more realistic here.

File: spec/features/search_page_spec.rb [hunk 7]
```diff
@@ -121,165 +79,88 @@
...
-                                             )
-                                             .build
-
-    index(
-      ministerial_dataset,
-      non_ministerial_dataset,
-      local_council_dataset,
-      non_departmental_public_body_dataset
-    )
+    dataset1 = build :dataset, organisation: build(:organisation, :raw)
```
Comment (benthorner) [inline spec/features/search_page_spec.rb]: Note: this test is actually passing by chance. The organisation boost functionality is broken.

File: spec/features/search_page_spec.rb [hunk 8]
```diff
@@ -121,165 +79,88 @@
...
-                                     .or eq('Forestry Commission')
-
-    expect(publishers[1]).to eq('Department for Environment Food & Rural Affairs')
-                               .or eq('English Heritage')
-                                     .or eq('Forestry Commission')
-
-    expect(publishers[2]).to eq('Department for Environment Food & Rural Affairs')
-                               .or eq('English Heritage')
-                                     .or eq('Forestry Commission')
+    publishers[0..2].each do |publisher|
```
Comment (benthorner) [inline spec/features/search_page_spec.rb]: This is a no-brainer instead of changing something 3 times.

File: spec/requests/error_handling_spec.rb [hunk 9]
```diff
@@ -15,20 +15,10 @@
   end
 
   it "handles a Datafile that can't be found by UUID" do
-    dataset_name = 'my-dataset'
-    dataset_uuid = SecureRandom.uuid
-    datafile_uuid = SecureRandom.uuid
```
Comment (benthorner) [inline spec/requests/error_handling_spec.rb]: We could leave this variable in, but it wouldn't improve the diff.

**Comment by benthorner** [comment]
The unnecessary changes should now be reverted and the diff a bit shorter. I've written a comment to justify each additional break from the original tests - they all made the change easier.

File: spec/features/search_page_spec.rb (hunk 7 above)
Comment (benthorner) [reply spec/features/search_page_spec.rb]: If you change the order the datasets are indexed in, the test fails.

File: spec/features/dataset_show_page_spec.rb [hunk 10]
```diff
@@ -101,385 +61,229 @@
...
       expect(page).to have_content('Topic: Not added')
     end
 
     scenario 'Last Updated field displays public_updated_at' do
-      dataset = DatasetBuilder.new.build
+      dataset = build :dataset
+      time = Time.parse(dataset.public_updated_at)
       index_and_visit(dataset)
-      expect(page).to have_content("Last updated: #{dataset['public_updated_at']}")
+      expect(page).to have_content("Last updated: #{time.strftime('%d %B %Y')}")
```
Comment (benthorner) [inline spec/features/dataset_show_page_spec.rb]: The builder was previously injecting an unrealistic (non-iso8601) string.

File: spec/features/dataset_show_page_spec.rb [hunk 11]
```diff
@@ -101,385 +61,229 @@
...
-        .build
-
-      index_and_visit(dataset)
-
-      expect(page).to have_css('h2', text: 'Data links')
-    end
   end
 
   feature 'Side bar content' do
+    let(:dataset1) { build :dataset, title: '1 Data Set' }
```
Comment (benthorner) [inline spec/features/dataset_show_page_spec.rb]: These datasets are common to all of the following scenarios, so it makes sense to extract them here instead of typing them over again. None of these scenarios require the dataset to have a datafile.

File: spec/features/dataset_show_page_spec.rb [hunk 12]
```diff
@@ -122,24 +99,19 @@
...
-                  .with_licence_custom('For feature specs only.')
-                  .build
+      dataset = build :dataset, :with_custom_licence,
+                                licence_custom: 'For feature specs only.'
 
       index_and_visit(dataset)
 
       expect(page)
-        .to have_css('meta[name="dc:rights"][content="Feature Spec Open License 2.1"]',
+        .to have_css('meta[name="dc:rights"][content="Example Open License 1.1"]',
```
Comment (benthorner) [inline spec/features/dataset_show_page_spec.rb]: The intention here is to re-use the factory trait for explicit/custom licences.
